pkg/iam/iam.go

```diff
@@ -0,0 +1,327 @@
+package iam
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	awsclient "github.com/aws/aws-sdk-go/aws/client"
+	awsiam "github.com/aws/aws-sdk-go/service/iam"
+	"github.com/go-logr/logr"
+)
+
+const (
+	ControlPlaneRole = "control-plane"
+	NodesRole        = "nodes"
+
+	IAMControllerOwnedTag = "capi-iam-controller/owned"
+	ClusterIDTag          = "sigs.k8s.io/cluster-api-provider-aws/cluster/%s"
+)
+
+type IAMServiceConfig struct {
+	AWSSession  awsclient.ConfigProvider
+	ClusterName string
+	IAMRoleName string
+	Log         logr.Logger
+	RoleType    string
+}
+
+type IAMService struct {
+	clusterName string
+	iamClient   *awsiam.IAM
+	iamRoleName string
+	log         logr.Logger
+	region      string
+	roleType    string
+}
+
+func New(config IAMServiceConfig) (*IAMService, error) {
+	if config.AWSSession == nil {
+		return nil, errors.New("cannot create IAMService with AWSSession equal to nil")
+	}
+	if config.ClusterName == "" {
+		return nil, errors.New("cannot create IAMService with empty ClusterName")
+	}
+	if config.IAMRoleName == "" {
+		return nil, errors.New("cannot create IAMService with empty IAMRoleName")
+	}
+	if config.Log == nil {
+		return nil, errors.New("cannot create IAMService with Log equal to nil")
+	}
+	if !(config.RoleType == ControlPlaneRole || config.RoleType == NodesRole) {
+		return nil, fmt.Errorf("cannot create IAMService with invalid RoleType '%s'", config.RoleType)
+	}
+	client := awsiam.New(config.AWSSession)
+
+	l := config.Log.WithValues("clusterName", config.ClusterName, "iam-role", config.RoleType)
+
+	s := &IAMService{
+		clusterName: config.ClusterName,
+		iamClient:   client,
+		iamRoleName: config.IAMRoleName,
+		log:         l,
+		roleType:    config.RoleType,
+		region:      client.SigningRegion,
+	}
+
+	return s, nil
+}
+
+func (s *IAMService) Reconcile() error {
+	s.log.Info("reconciling AWSMachineTemplate")
```

Comment:
Since this could also be coming from the `AWSMachinePool` later it may be better to say "Reconciling AWS IAM role" or similar, WDYT?

Reply:
good point